File: C++/binary_tree_paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <string> 

class Solution {
public:
    void storePaths(vector<string>& result, TreeNode* root, string path) {
            if(root == NULL) return;

            if(root->right == NULL && root->left == NULL) {
                result.push_back(path + to_string(root->val));
                return;
            }
            
            path += to_string(root->val) + "->";
            
            storePaths(result, root->left, path);
            storePaths(result, root->right, path);
    }
    
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> result;
        storePaths(result, root, "");      
        return result;
    }
};

File: Python3/binary_tree_paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        def storePaths(root, path):
            if not root: return None

            if not root.right and not root.left:
                result.append(path + str(root.val))
                return None
            
            path += str(root.val) + "->"
            
            storePaths(root.left, path)
            storePaths(root.right, path)
        
        result = []
        storePaths(root, "")
        
        return result
